trin-core/src/utp/packets.rs

```diff
@@ -0,0 +1,322 @@
+pub const HEADER_SIZE: usize = 20;
+pub const VERSION: u8 = 1;
+
+#[derive(PartialEq, Debug)]
+pub enum PacketType {
+    Data,
+    Fin,
+    State,
+    Reset,
+    Syn,
+}
+
+impl TryFrom<u8> for PacketType {
+    type Error = &'static str;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        match value {
+            0 => Ok(PacketType::Data),
+            1 => Ok(PacketType::Fin),
+            2 => Ok(PacketType::State),
+            3 => Ok(PacketType::Reset),
+            4 => Ok(PacketType::Syn),
+            _ => Err("invalid packet type"),
+        }
+    }
+}
+
+impl From<PacketType> for u8 {
+    fn from(value: PacketType) -> u8 {
+        match value {
+            PacketType::Data => 0,
+            PacketType::Fin => 1,
+            PacketType::State => 2,
+            PacketType::Reset => 3,
+            PacketType::Syn => 4,
+        }
+    }
+}
+
+impl<'a> TryFrom<&'a [u8]> for PacketHeader {
+    type Error = &'static str;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() < HEADER_SIZE {
+            return Err("invalid header size");
+        }
+
+        if value[0] & 0xf != VERSION {
+            return Err("invalid packet version");
+        }
+
+        if let Err(e) = PacketType::try_from(value[0] >> 4) {
+            return Err(e);
+        }
+
+        Ok(PacketHeader::decode(value))
+    }
+}
+
+fn check_extensions(data: &[u8]) -> Result<(), &'static str> {
+    if data.len() < HEADER_SIZE {
+        return Err("invalid header size");
+    }
+
+    let mut extension = data[1];
+    let mut index = HEADER_SIZE;
+
+    // must be at least 4 bytes, and in multiples of 4
+    while extension != 0 {
+        if data.len() < index + 2 {
+            return Err("Invalid packet length");
+        }
+        extension = data[index];
+        let len = data[index + 1] as usize;
+
+        if len == 0 || len % 4 != 0 || len + index + 2 > data.len() {
+            return Err("Invalid Extension Length");
+        }
+
+        index += len + 2;
+    }
+    Ok(())
+}
+
+impl<'a> TryFrom<&'a [u8]> for Packet {
+    type Error = &'static str;
+
+    // todo: Added support to check validity of extensions,
+    // we don't need to worry about checking payloads
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        match PacketHeader::try_from(value).and(check_extensions(value)) {
+            Ok(_) => Ok(Packet(value.to_owned())),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl Default for PacketHeader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct PacketHeader {
+    type_ver: u8,
+    extension: u8,
+    connection_id: u16,
+    // This time is in microseconds
+    timestamp: u32,
```

Comment:
A 32-bit timestamp in microseconds can only count up to ~4200 seconds, so this could really benefit from an explanation about why rollover isn't a concern.

Reply:
The actual time doesn't matter this is just for congestion control. Which is generated from get_time()

Reply:
Ah yes, I see what's happening now in `get_time()`. Ok, well I will create a separate issue on this discussion, since this PR is just a refactor and doesn't really touch the logic about timestamps. Key issues:
- Calling that thing a timestamp is very confusing. It's a reference counter that rolls over every ~hour, which just feels very different. So maybe we should name it `usec_counter` or something like that.
- Calculating the time difference after a rollover looks wrong to me. Current time (rolled over) minus previous time (before rollover) is not equal to `previous_time - current_time`. Instead it's `U32_MAX - previous_time + current_time`.

Reply:
That is what it is called in the spec. Also to be fair it still is technically a timestamp

Reply:
Also, I  moved the refactor to #226 with some other changes because this PR is large enough and it isn't really related to this PR.

Reply:
> Also to be fair it still is technically a timestamp

I disagree, but #227 seems like the right place to hash that out. :)